In [ ]:

# Create a synthetic step signal representing gaze segments.
import numpy as np
from pymovements.events.detection import ihmm
from pymovements.gaze.transforms_numpy import pos2vel
from pymovements.synthetic import step_function
from pymovements.gaze import from_numpy
positions = step_function(
         length=200, steps=[2, 5, 9, 111, 150],
         values=[(1., 2.), (2., 3.), (3., 4.), (1., 1.), (2., 2.)],
         start_value=(0., 0.),
     )
#positions.shape

# Transform into velocities

velocities = pos2vel(positions)

#velocities.shape

#    Apply event detection algorithm on numpy array:

ihmm(velocities)

#    Run fixation detection with custom HMM parameters:

dict = {'mu': [2.0140785987072225, 69.41529375180251], 'sigma': [1.3220152347857494, 87.32409626093246], 'init': [1.e+00, 1.e-12], 'trans': [[0.97360507, 0.02639493],
       [0.07593547, 0.92406453]]}

ihmm(velocities, hmm_parameters_dict = dict)


#    Polars series are also supported as input. Let's create a nested position series from our numpy
#    array:

df = polars.from_numpy(velocities, schema=['x', 'y'])
velocities_series = df.select(polars.concat_list(('x', 'y')).alias('velocity'))['velocity']
velocities_series
   

#    Apply event detection algorithm on polars series:

ihmm(velocities_series)
    

#    We can also apply the detection on a :py:class:`~pymovements.Gaze` object.

from pymovements import Experiment
gaze = from_numpy(
        velocity=velocities.T,
        time=np.arange(len(velocities)),)
gaze


#    Run fixation detection by using the :py:meth:`~pymovements.Gaze.detect` method.

gaze.detect('ihmm')
gaze.events


#    Passing parameters to :py:meth:`~pymovements.Gaze.detect`:

gaze.detect('idt', reestimation=True, hmm_parameters_dict = dict, name='fixation_ihmm')
gaze.events.filter_by_name('fixation_idt')


In [1]:
import numpy as np
from pymovements.events.detection import ihmm
from pymovements.gaze.transforms_numpy import pos2vel
from pymovements.synthetic import step_function
from pymovements.gaze import from_numpy
positions = step_function(
         length=200, steps=[2, 5, 9, 111, 150],
         values=[(1., 2.), (2., 3.), (3., 4.), (1., 1.), (2., 2.)],
         start_value=(0., 0.),
     )
positions.shape

(200, 2)

In [2]:
# Transform into velocities

velocities = pos2vel(positions)

velocities.shape

(200, 2)

In [3]:
#   Apply event detection algorithm on numpy array:

ihmm(velocities)

name,onset,offset,duration
str,i64,i64,i64
"""fixation""",11,108,97
"""fixation""",113,147,34
"""fixation""",152,199,47


In [4]:
#    Run fixation detection with custom HMM parameters:

dict = {'mu': [2.0140785987072225, 69.41529375180251], 'sigma': [1.3220152347857494, 87.32409626093246], 'init': [1.e+00, 1.e-12], 'trans': [[0.97360507, 0.02639493],
       [0.07593547, 0.92406453]]}

ihmm(velocities, hmm_parameters_dict = dict)

name,onset,offset,duration
str,i64,i64,i64
"""fixation""",0,0,0
"""fixation""",11,108,97
"""fixation""",113,147,34
"""fixation""",152,199,47


    ┌──────────┬───────┬────────┬──────────┐
    │ name     ┆ onset ┆ offset ┆ duration │
    │ ---      ┆ ---   ┆ ---    ┆ ---      │
    │ str      ┆ i64   ┆ i64    ┆ i64      │
    ╞══════════╪═══════╪════════╪══════════╡
    │ fixation ┆ 0     ┆ 0      ┆ 0        │
    │ fixation ┆ 11    ┆ 108    ┆ 97       │
    │ fixation ┆ 113   ┆ 147    ┆ 34       │
    | fixation ┆ 152   ┆ 199    ┆ 47       |
    └──────────┴───────┴────────┴──────────┘

    
"fixation"	0	0	0
"fixation"	11	108	97
"fixation"	113	147	34
"fixation"	152	199	47


In [6]:
from pymovements import Experiment
gaze = from_numpy(
        position=velocities.T,
        time=np.arange(len(velocities)),
     )
gaze

time,position
i64,list[f64]
0,"[0.0, 0.0]"
1,"[500.0, 1000.0]"
2,"[333.333333, 666.666667]"
3,"[333.333333, 500.0]"
4,"[333.333333, 333.333333]"
…,…
195,"[0.0, 0.0]"
196,"[0.0, 0.0]"
197,"[0.0, 0.0]"
